zerver/actions/user_topics.py

```diff
@@ -55,3 +56,45 @@ def do_set_user_topic_visibility_policy(
     }
 
     send_event(user_profile.realm, user_topic_event, [user_profile.id])
+
+
+def bulk_do_set_user_topic_visibility_policy(
+    user_profiles: List[UserProfile],
+    stream: Stream,
+    topic: str,
+    *,
+    visibility_policy: int,
+    last_updated: Optional[datetime.datetime] = None,
+    skip_muted_topics_event: bool = False,
+) -> None:
+    if last_updated is None:
+        last_updated = timezone_now()
+
+    bulk_set_user_topic_visibility_policy_in_database(
+        user_profiles,
+        stream.id,
+        topic,
+        visibility_policy=visibility_policy,
+        recipient_id=stream.recipient_id,
+        last_updated=last_updated,
+    )
+
+    for user_profile in user_profiles:
+        # This first muted_topics event is deprecated and will be removed
+        # once clients are migrated to handle the user_topic event type
+        # instead.
+        if not skip_muted_topics_event:
+            muted_topics_event = dict(
+                type="muted_topics", muted_topics=get_topic_mutes(user_profile)
+            )
+            send_event(user_profile.realm, muted_topics_event, [user_profile.id])
+
+        user_topic_event: Dict[str, Any] = {
+            "type": "user_topic",
+            "stream_id": stream.id,
+            "topic_name": topic,
+            "last_updated": datetime_to_timestamp(last_updated),
+            "visibility_policy": visibility_policy,
+        }
+
+        send_event(user_profile.realm, user_topic_event, [user_profile.id])
```

Comment:
This code seems same as that in the non-bulk function. It might be worth it to extract out a common function to just send the events correctly in a prep commit, and then use that function here. (So that we have a single place to handle the compatibility logic.)

Reply:
Yeah that'd be a huge help; we often name such extracted functions something like `notify_user_topics`

Reply:
Since the updated approach is not to maintain two parallel functions (do_set_ & bulk_do_set_ ). We no longer need this.